### PR TITLE
Minor tweaks plus fixing #55 & #25

### LINKED ISSUE #25: Signals are not handled correctly
Hi there,

when running a hello-world app from the Docker image, the Unix signals are not handled properly.
Running Payara in foreground and <kbd>Ctrl</kbd>+<kbd>C</kbd> doesn't stop the server. Similarly, running `docker stop` takes ten seconds -- the default wait time, before Docker sends a SIGKILL.

See https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/

Tested with `payara/server-full:5-SNAPSHOT` and `payara/server-full:latest` (as of today).

### LINKED ISSUE #55: asadmin command not found
At least with payara/server-full:latest@sha256:9ba625667910d775a40e47100730a320cbf59e0f2732e39c1262e3ebd74e9f18 asadmin is not on path and it's required to first navigate to /opt/payara5/bin

### PR DESCRIPTION
**Pull request by poikilotherm**
This PR adds:

* A few minor tweaks to the Dockerfile
* Add the JVM option to have proper heap spacing around
* Fixes #55
* Fixes #25 by providing direct tini support
* Adds options for more complex init scripts at startup time and uses it for the deploy command generation. Possible use cases: commonly known `wait_for_it.sh`, load data, complex Glassfish commands, ...

I am happy to change things on your request. Also allowing direct pushes to my feature branch for all upstream maintainers.

Feedback is very much appreciated.

Cheers,
@poikilotherm

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -25,25 +28,30 @@ ENV HOME_DIR=/opt/payara\
     DEPLOY_PROPS=\
     POSTBOOT_COMMANDS=/opt/payara/config/post-boot-commands.asadmin\
     PREBOOT_COMMANDS=/opt/payara/config/pre-boot-commands.asadmin
+ENV PATH="${PATH}:${PAYARA_DIR}/bin"
```
Comment (shoeper) [inline Dockerfile]: can't it be merged with the previous env?

File: Dockerfile [hunk 2]
```diff
@@ -25,25 +28,30 @@ ENV HOME_DIR=/opt/payara\
...
-RUN groupadd payara && \
-    useradd -b ${HOME_DIR} -M -s /bin/bash -d ${HOME_DIR} payara -g payara && \
+RUN groupadd -g 1000 payara && \
+    useradd -u 1000 -M -s /bin/bash -d ${HOME_DIR} payara -g payara && \
     echo payara:payara | chpasswd && \
     mkdir -p ${DEPLOY_DIR} && \
     mkdir -p ${CONFIG_DIR} && \
     mkdir -p ${SCRIPT_DIR} && \
-    chown -R payara:payara ${HOME_DIR}
+    chown -R payara. ${HOME_DIR}
```
Comment (shoeper) [inline Dockerfile]: I think the previous version was more readable and also don't see the . syntax (?) documented in the chown man page.

File: Dockerfile [hunk 3]
```diff
@@ -25,25 +28,30 @@ ENV HOME_DIR=/opt/payara\
...
+
+# Install tini as minimized init system
+ADD https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini /tini
```
Comment (shoeper) [inline Dockerfile]: Adding and checking a checksum would make it more secure. But as I see it is also missing for the payara download which is even done over http.

File: Dockerfile [hunk 4]
```diff
@@ -7,15 +7,18 @@ FROM openjdk:8u171-jdk
 # 8181: https
 EXPOSE 4848 9009 8080 8181
 
+# Payara version (5.183+)
+ARG PAYARA_VERSION=5.183
+ARG PAYARA_PKG=http://central.maven.org/maven2/fish/payara/distributions/payara/${PAYARA_VERSION}/payara-${PAYARA_VERSION}.zip
```
Comment (shoeper) [inline Dockerfile]: Not really part of this PR but fetching it over https would be more secure and checking a checksum even better.

https://repo1.maven.org/maven2/fish/payara/distributions/payara/5.183/payara-5.183.zip

https://search.maven.org/remotecontent?filepath=fish/payara/distributions/payara/5.183/payara-5.183.zip

File: Dockerfile [hunk 5]
```diff
@@ -7,15 +7,18 @@ FROM openjdk:8u171-jdk
...
+ARG TINI_VERSION=v0.18.0
```
Comment (MattGill98) [inline Dockerfile]: I don't understand the need for this to provide 3 separate layers to the environment variable block below?

File: Dockerfile (hunk 2 above)
Comment (MattGill98) [inline Dockerfile]: Could you explain this change?

File: Dockerfile (hunk 5 above)
Comment (poikilotherm) [reply Dockerfile]: IMHO using an ARG for build only variables is a better approach.
ARG should no longer create layers according to https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers

File: Dockerfile (hunk 2 above)
Comment (poikilotherm) [reply Dockerfile]: Sure - as explained in the git commit message, this is a usefull shortcut. If the group would be changed someday, you do not need to update the statement.

See also: https://linux.die.net/man/1/chown
> "If a colon but no group name follows the user name, that user is made the owner of the files and the group of the files is changed to that user's login group."

File: Dockerfile (hunk 1 above)
Comment (poikilotherm) [reply Dockerfile]: Nope. You cannot re-use an ENV var you just created within the same ENV. Already tried and failed...

File: Dockerfile (hunk 2 above)
Comment (poikilotherm) [reply Dockerfile]: Please see my [comment above](https://github.com/payara/docker-payaraserver-full/pull/61#discussion_r235741106).

Comment (shoeper) [reply Dockerfile]: But it is a dot??, not a colon ...

File: Dockerfile (hunk 3 above)
Comment (poikilotherm) [reply Dockerfile]: Sure. The tini project offers GPG signatures more for those checks, but this needs more stuff within the image. We could go for a multistage build to avoid artifacts in the resulting image.

File: Dockerfile (hunk 2 above)
Comment (poikilotherm) [reply Dockerfile]: Right :smile: 

Had to dig a bit deeper:
http://www.gnu.org/software/coreutils/manual/html_node/chown-invocation.html
> Some older scripts may still use ‘.’ in place of the ‘:’ separator. POSIX 1003.1-2001 (see Standards conformance) does not require support for that, but for backward compatibility GNU chown supports ‘.’ so long as no ambiguity results. New scripts should avoid the use of ‘.’ because it is not portable, and because it has undesirable results if the entire owner‘.’group happens to identify a user whose name contains ‘.’. 

That was news to me, learned about the "." during LPIC seminars 4 years ago. Will change this to a colon.

File: Dockerfile (hunk 4 above)
Comment (OndrejM) [reply Dockerfile]: Validating the download with a checksum would be useful. And I agree with using HTTPS.

@MattGill98, what do you think?

File: Dockerfile (hunk 1 above)
Comment (shoeper) [reply Dockerfile]: Thanks

File: Dockerfile (hunk 3 above)
Comment (OndrejM) [reply Dockerfile]: I'm giving +1.

File: Dockerfile (hunk 4 above)
Comment (OndrejM) [reply Dockerfile]: Looks good. Is it OK just to download the SHA1 checksum from Maven and insert it into the PAYARA_SHA1 arg for future releases? We'll need to document that step in our release guide.

**Comment by poikilotherm** [comment]
While driving home, sth. came to my mind...

What do you guys prefer: the init script extension as I did it now, or use a subdirectory that can be used as a volume target? Using with a volume makes `docker-compose` usage easier without creating another image.

File: Dockerfile (hunk 4 above)
Comment (poikilotherm) [reply Dockerfile]: I did it just like that. :wink: You can of course calculate it yourself via `sha1sum` and not trust the Maven repo. (At least I don't see any other point in using the checksum than this bit of paranoia...)

File: Dockerfile (hunk 3 above)
Comment (poikilotherm) [reply Dockerfile]: Done in b8b1ac0.

File: Dockerfile (hunk 2 above)
Comment (poikilotherm) [reply Dockerfile]: @MattGill98 can this be resolved?

File: Dockerfile (hunk 4 above)
Comment (shoeper) [reply Dockerfile]: Okay, you already edited it.

The reason I see is that someone could replace the artifact on maven. I'm not sure how it works in  detail with maven repo, but account hijacking happens from time to time.

File: Dockerfile (hunk 5 above)
Comment (MattGill98) [reply Dockerfile]: Good spot, if the ARG statement doesn't make layers this makes a lot more sense!

File: Dockerfile (hunk 2 above)
Comment (MattGill98) [reply Dockerfile]: Thank you for your clarification @poikilotherm, I'm happy with this change.

File: Dockerfile (hunk 3 above)
Comment (MattGill98) [reply Dockerfile]: @poikilotherm I like the GPG signing. Having said that, would it be possible to replace the ADD with a curl to simplify this to one single layer? If the ADD provides any extra value I'm happy to leave this as is!

File: Dockerfile (hunk 4 above)
Comment (MattGill98) [reply Dockerfile]: I'm happy with this change, looks good!

**Comment by MattGill98** [comment]
Hi @poikilotherm,

What exactly do you mean by the second option? I don't quite see how volume mounting would provide an alternative to the init extension.

Kind regards,

Matt

**Comment by poikilotherm** [comment]
Sorry, maybe it was a bit late yesterday evening...

An example can be found in the [entrypoint of MySQL image](https://github.com/docker-library/mysql/blob/9e0ea0bed1e0740a733d224b893b7b8fcd692635/5.7/docker-entrypoint.sh#L196).

They execute scripts and more from a directory at startup, which can be filled using a volume mapping without the need to create a sub-image. With my current approach, you have to create a sub-image.

**Comment by poikilotherm** [comment]
Alright, done in 32fabe1

This can even be extendend to support sth. like `*.asadmin` or similar appended to the boot scripts on startup.

File: Dockerfile (hunk 3 above)
Comment (poikilotherm) [reply Dockerfile]: This is just copy & paste from the tini project. I could of course put all of this in the RUN that is executed after the ADD.

Comment (poikilotherm) [reply Dockerfile]: Done in 8f06c50.
I think this can be resolved now, correct?

File: bin/entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+# Run init scripts (credits go to MySQL Docker entrypoint script)
+for f in ${SCRIPT_DIR}/init_* ${SCRIPT_DIR}/init.d/*; do
```
Comment (MattGill98) [inline bin/entrypoint.sh]: Does this line list the files in an aggregated sort order, or will the `init_` scripts all be ran first?

Comment (poikilotherm) [reply bin/entrypoint.sh]: Init scripts will run first. IMHO it is unlikely you will use both options at once...

Comment (MattGill98) [reply bin/entrypoint.sh]: I wouldn't suggest a sort, rather just that it might be worth mentioning in the README!

Comment (poikilotherm) [reply bin/entrypoint.sh]: [Noted in README](https://github.com/payara/docker-payaraserver-full/pull/61/commits/8089f010a859637eddd34f6716eeb6aec1ba0c85#diff-04c6e90faac2675aa89e2176d2eec7d8R109)

Comment (poikilotherm) [reply bin/entrypoint.sh]: @MattGill98 I added it on your request :smile:

**Comment by poikilotherm** [comment]
Thx Matt :smile_cat: 

Crossing finger this is merged and released soon :wink:

**Comment by poikilotherm** [comment]
@MattGill98 thank you very much for merging :smile:

**Comment by MattGill98** [comment]
Thank you very much for the contribution! These changes will be available in the 5.184 docker image :smile:
